Реализация деплоя фронтенда, создание SPA

geogreck opened the pull request:
closes #58

geogreck committed:
Обновлен package.json

geogreck commented:
Произошел отвал `HEAD`.

geogreck committed:
Добавлен самодельный `static` сервер

geogreck committed:
Добавлена реализация `SPA`

geogreck committed:
Обновлены параметры сборки:
 - В `Dockerfile` добавлен новый `target`
 - `docker-compose` разделен на `prod` и `dev` версии
 -  В `Makefile` обновлены старые команды, добавлены новые для деплоя

geogreck committed:
Обновлены скрипты `package.json`

geogreck committed:
Обновлена документация

geogreck commented:
# **TO-DO**

- [x] Пофиксить ЛИНТЕР
- [ ] Убедиться, что деплоймент `dev` версии не пострадал, работоспобность `Makefile` полностью сохранена
- [x] Улучшить Handler в `static` сервере
- [x] Произвести рефакторинг деплоймент скриптов (Создание хеш таблицы путей)
- [x] Скрипты роутера портировать на `TS`(?)

geogreck committed:
Рефакторинг скриптов `SPA`

geogreck committed:
Изменен тип модулей JS

geogreck committed:
Merge branch 'main' into frontend-cd

geogreck committed:
Инвентарь добавлен в список страниц `SPA`

geogreck committed:
Содержимое вкладки теперь обновляется при переходе

geogreck committed:
Merge branch 'main' into frontend-cd

geogreck committed:
Фикс ЕГО

geogreck committed:
Повышение отказоустойчвости
todolist'a

geogreck committed:
Добавлено выделение раздела по умолчанию

geogreck committed:
Добавлена поддержка старого режима `yarn dev`

geogreck committed:
Починены классы html базы

geogreck committed:
Добавлен скрипт профиля

geogreck committed:
Добавлено "пособие" по разработке фронтенда

geogreck committed:
Починил форматирование

geogreck committed:
Рефакторинг `SPA` скриптов

geogreck committed:
Eslint настроен для работы с `NodeJS`

geogreck committed:
Улучшено логирование запросов

geogreck committed:
Добавлена подмена стартовой страницы в `prod`

geogreck committed:
Запуск с подменой `start_page` добавлен в Docker

geogreck committed:
Рефакторинг деплой скриптов

stewkk commented on Makefile:
```diff
@@ -2,11 +2,11 @@ all: dev
 dev:
-	@docker compose up --build
+	@docker compose -f docker-compose.dev.yml up --build
 
```
Отдельно фронт есть в режиме дев и прод, а все контейнеры только в дев. Мб, стоит добавить прод для всех контейнеров:
```suggestion

.PHONY: prod
prod:
	@docker compose up --build
	
```

stewkk replied:
И, кстати, возможно стоит сделать `prod` режимом по умолчанию (при запуске с помощью `make`), т.к. сейчас `docker compose up` поднимает прод версию.

stewkk commented on frontend/deploy/server.js:
```diff
@@ -0,0 +1,186 @@
+                    req.url,
+                    ' ',
+                    code === '404' ? '\x1b[31m' : '\x1b[32m',
+                    code,
```
Нужны константы для цветов

stewkk commented on frontend/deploy/index.js:
```diff
@@ -0,0 +1,15 @@
+    })
+} catch (error) {
+    console.error(error)
+    throw new error()
```
Кринжанул =)

stewkk commented on frontend/guide.md:
````diff
@@ -0,0 +1,40 @@
+  - [ ] Мой код имеет `onLoad` функцию, которая получает с сервера данные и отображает их на странице ([Пример](js/list/todolist.ts#L18))
+  - [ ] Мой код работает как и в `SPA`, так и в `dev` режиме. Например, добавлена волшебная конструкция вида:
+    ```js
+        try {
+            onTodoLoad()
+        } catch (error) {}
````
😂

stewkk commented on frontend/html/spa.html:
```diff
@@ -0,0 +1,163 @@
+                </div>
+              </body>
+
+            </html>
```
```suggestion
            </html>

```

stewkk replied:
И немного поехало форматирование =)

geogreck replied:
Не немного.

stewkk commented on frontend/js/inventory/inventory.ts:
```diff
@@ -13,25 +13,33 @@ let Equipped = {
+    modalInventory = new bootstrap.Modal(<HTMLFormElement>document.getElementById('inventoryModal'))
+    // получение предметов с сервера
+    sendRequest('GET', server + '/items').then((data) => {
+        for (let i = 0; i < data.length; i++) {
```
Не канон!

stewkk commented on frontend/js/inventory/inventory.ts:
```diff
@@ -13,25 +13,33 @@ let Equipped = {
+            let userImg = document.getElementById('inventory__user-img')
+            let id = Equipped.skin
+            let item = itemsInventory.get(id)!
+            //userImg!.setAttribute('src', `https://wg.grechkogv.ru/assets/${item.imageSrc}`)
```
Страшно

stewkk commented on frontend/js/list/todolist.ts:
```diff
@@ -9,30 +9,51 @@ var labels = new Map<number, Label>()
+            let buf = data[i]
+            let newTask = new Task(buf.id, buf.name, buf.description, buf.dueDate, buf.labels, false)
+            newTask.setStatus(buf.status, false)
+            tasks.set(buf.id, newTask)
```
Печальная инициализация :/

stewkk commented on frontend/js/spa/route.ts:
```diff
@@ -0,0 +1,43 @@
+    private name: string
+    private htmlName: string
+    private isDefault: boolean
+    private fn: Function
```
Плохое название. Непонятно, что делает эта функция, и почему `evalFn` вызывает ее таким странным образом.

stewkk commented on frontend/js/spa/route.ts:
```diff
@@ -0,0 +1,43 @@
+        return this.isDefault
+    }
+
+    public evalFn() {
+        let count = 0
+        let timer = setInterval(() => {
+            this.fn()
+            count++
+            if (document.getElementById('modal') || document.getElementById('inventoryModal') || count > 10) {
+                clearInterval(timer)
+            }
+        }, 100)
```
Хочется плакать

geogreck replied:
Мне тоже

stewkk reviewed (changes requested):
> Как хорошо, что я не фронтендер!

P.S. Поправь варнинги линтера =)

geogreck committed:
Александр направил меня на путь истинный
- По умолчанию теперь запускается prod режим
- Для цветов сделана отдельная хеш-таблица
- Убран код, отмеченный в ревью, как "кринжанул"
- Отформатирован html файл
- убран цикл for в загрузке инвентаря
- поле fn переименовано в onLoadFn
- улучшен код, оцененный как "хочется плакать"

И ещё мое личное прозрение
- внутрь SPA добавлена своя страница 404
- в целом улучшена функция направления на страницу

Co-authored-by: Alexandr Starovoytov <43303621+stewkk@users.noreply.github.com>

geogreck commented:
Забыл добавить, что у страниц теперь наконец-то есть title!

geogreck committed:
Хеш таблица цветов вынесена в отдельный файл
Так же улучшил JSDoc

geogreck committed:
Названия страниц теперь красивые и на русском
